src/Collection/Iterator/UnfoldIterator.php

```diff
@@ -57,34 +47,76 @@ class UnfoldIterator extends IteratorIterator implements RecursiveIterator
      */
     public function __construct(Traversable $items, callable $unfolder)
     {
-        $this->_unfolder = $unfolder;
-        parent::__construct($items);
-        $this->_innerIterator = $this->getInnerIterator();
+        $itemsIterator = new class ($items, $unfolder) extends IteratorIterator implements RecursiveIterator {
+            /**
+             * @var callable
+             */
+            protected $unfolder;
+
+            /**
+             * Creates the iterator that will generate child iterators from each of the
+             * elements it was constructed with.
+             *
+             * @param \Traversable $items The list of values to iterate
+             * @param callable $unfolder A callable function that will receive the
+             * current item and key. It must return an array or Traversable object
+             * out of which the nested iterators will be yielded.
+             */
+            public function __construct(Traversable $items, callable $unfolder)
+            {
+                parent::__construct($items);
+                $this->unfolder = $unfolder;
+            }
+
+            /**
+             * @return bool
+             */
+            public function hasChildren(): bool
+            {
+                return true;
+            }
+
+            /**
+             * @return \RecursiveIterator
+             */
+            public function getChildren(): RecursiveIterator
+            {
+                return new NoChildrenIterator(($this->unfolder)($this->current(), $this->key()));
+            }
+        };
+
+        parent::__construct(new RecursiveIteratorIterator($itemsIterator));
     }
 
     /**
-     * Returns true as each of the elements in the array represent a
-     * list of items
+     * See \IteratorIterator::key)
      *
-     * @return bool
+     * @return mixed
      */
-    public function hasChildren(): bool
+    public function key(): mixed
```

Comment:
```suggestion
    public function key()
```
Can't used `mixed` in 4.x.